Source/Core/Common/WorkQueueThread.h

```diff
@@ -70,6 +70,19 @@ class WorkQueueThread
     }
   }
 
+  // Doesn't return until the most recent function invocation has finished.
+  void Flush()
+  {
+    if (m_thread.joinable())
+    {
+      m_flush.Set();
+      Clear();
```

Comment:
Why do you explicitly clear the working queue? It should be empty after the wait call anyway and this throws the allocation away, doesnt it?

Reply:
Yes... This is hyer-specialised for my usecase, where I actually want "AbortAndFlush". 

Because each work item takes 20ms minimum, I wanted the queue to empty as soon as possible. I should fix it to actually do what you expect a function named flush to do, and just manually call `Clear()` in my code. 

> This also deadlocks if you call it when the thread is idle as far as I can tell. (Flush(); Flush();)

Yes... might be hard to trigger because the Clear function does a wakeup. But there is a race condition if the thread sleeps between `Clear()` and `m_flushed.Wait();`

Reply:
https://github.com/dolphin-emu/dolphin/pull/11417/commits/5d86c88edb867955a813d978ce03b480d1ffb8c3

Any thoughts on this?

Reply:
Um.... 
It's missing the `is_flushing()` feature that I need [here](https://github.com/dolphin-emu/dolphin/pull/11473/files#diff-d3cfae5376cf8d3755cede57632578d528203cf113e8fa6d9546e3b05d9065bcR42). The code is about 20% longer. Harder to reason about correctness. And I can spot at least one bug that the Flag wrapper was designed to prevent.

I don't think it's an improvement.

Reply:
Mhm, probably just me then. I find the 4 different Flags + Events much harder to reason about.

Reply:
I don't know. Maybe I just find condition vars slightly hard to reason about.   
The advantage of the Flag and Event wrapper is that they are idiot-proof. Near impossible to mess up. I can look at the calls and know it's going to do the right thing. 

Your code essentially just replaced Flags with (atomic) bools and Events with condition vars. Which is what they both are internally.

Yes, your code is slightly more efficient. One you add a forth atomic bool to implement `is_flushing()` it manages to use two less mutexes and one less atomic bool than the Event/Flag version.

Reply:
> The advantage of the Flag and Event wrapper is that they are idiot-proof. Near impossible to mess up. I can look at the calls and know it's going to do the right thing.

With multiple flags and events you often rely on the exact order of those getting checked and cleared and waiting on an event. That isn't shared by a  of that is protected by a lock. That seems far from idiot-proof to me.

I'm pretty sure it also relies on a strict 1:1 relationship between the thread Flushing and the worker thread. If you have multiple threads interacting with the WorkQueueThread, it will probably fall apart.

That's just my 2 cents though. At the end of the day, I'm perfectly fine with keeping the current implementation (and adding a proper FlushAll + Push).

Reply:
Though we are getting a little side tracked with personal preferences about what's easier to reason about. 

The real question is what implementation we should go with, and you do make a good argument that your implementation is superior (I didn't even consider the multiple threads interacting usecase, because I wasn't planning to use it).

It might be worth going with your option (after adding the IsFlushing feature I need)